File: scripts/macos-transcribe.js
ObjC.import("Foundation");
ObjC.import("Speech");

function unwrap(value) {
  return ObjC.unwrap(value);
}

function isRealObjCValue(value) {
  return value !== undefined && value !== null && String(value) !== "[id nil]";
}

function resolveLocaleIdentifier(rawValue) {
  const locale = String(rawValue || "").trim();
  if (!locale || locale.toLowerCase() === "auto") {
    return unwrap($.NSLocale.autoupdatingCurrentLocale.localeIdentifier);
  }
  return locale;
}

function requestAuthorization() {
  const currentStatus = $.SFSpeechRecognizer.authorizationStatus;
  if (currentStatus === $.SFSpeechRecognizerAuthorizationStatusAuthorized) {
    return;
  }

  let done = false;
  let resolvedStatus = currentStatus;

  $.SFSpeechRecognizer.requestAuthorization(function (status) {
    resolvedStatus = status;
    done = true;
  });

  let attempts = 0;
  while (!done && attempts < 50) {
    $.NSRunLoop.currentRunLoop.runUntilDate($.NSDate.dateWithTimeIntervalSinceNow(0.2));
    attempts += 1;
  }

  if (resolvedStatus !== $.SFSpeechRecognizerAuthorizationStatusAuthorized) {
    throw new Error(
      "Speech recognition authorization was denied. Allow Obsidian or osascript in System Settings > Privacy & Security > Speech Recognition."
    );
  }
}

function transcribeAudioFile(audioPath, localeIdentifier) {
  const recognizerLocale = $.NSLocale.alloc.initWithLocaleIdentifier(localeIdentifier);
  const recognizer = $.SFSpeechRecognizer.alloc.initWithLocale(recognizerLocale);

  if (!recognizer || !recognizer.isAvailable) {
    throw new Error("Speech recognizer is unavailable for locale: " + localeIdentifier);
  }

  const url = $.NSURL.fileURLWithPath(audioPath);
  const request = $.SFSpeechURLRecognitionRequest.alloc.initWithURL(url);
  request.shouldReportPartialResults = false;

  let done = false;
  let transcript = "";
  let recognitionError = "";
  recognizer.recognitionTaskWithRequestResultHandler(
    request,
    function (result, error) {
      if (isRealObjCValue(error)) {
        recognitionError = unwrap(error.localizedDescription);
        done = true;
        return;
      }

      if (isRealObjCValue(result)) {
        transcript = unwrap(result.bestTranscription.formattedString);
        if (unwrap(result.isFinal)) {
          done = true;
        }
      }
    }
  );

  let attempts = 0;
  while (!done && attempts < 600) {
    $.NSRunLoop.currentRunLoop.runUntilDate($.NSDate.dateWithTimeIntervalSinceNow(0.2));
    attempts += 1;
  }

  if (!done) {
    throw new Error("Speech recognition timed out.");
  }

  if (recognitionError) {
    throw new Error(recognitionError);
  }

  transcript = String(transcript || "").trim();
  if (!transcript) {
    throw new Error("Speech recognizer returned an empty transcript.");
  }

  return transcript;
}

function run(argv) {
  if (!argv || argv.length < 1) {
    throw new Error("Usage: osascript -l JavaScript macos-transcribe.js <audio-path> [locale]");
  }

  const audioPath = argv[0];
  const localeIdentifier = resolveLocaleIdentifier(argv[1]);

  requestAuthorization();
  const transcript = transcribeAudioFile(audioPath, localeIdentifier);
  console.log(transcript);
}

File: scripts/macos-transcribe.swift
import Foundation
import Speech

enum CliError: LocalizedError {
    case usage
    case fileNotFound(String)
    case authorizationDenied
    case recognizerUnavailable(String)
    case emptyTranscript
    case timedOut

    var errorDescription: String? {
        switch self {
        case .usage:
            return "Usage: swift macos-transcribe.swift <audio-path> [locale]"
        case .fileNotFound(let path):
            return "Audio file not found: \(path)"
        case .authorizationDenied:
            return "Speech recognition authorization was denied. Allow Obsidian in System Settings > Privacy & Security > Speech Recognition."
        case .recognizerUnavailable(let locale):
            return "Speech recognizer is unavailable for locale: \(locale)"
        case .emptyTranscript:
            return "Speech recognizer returned an empty transcript."
        case .timedOut:
            return "Speech recognition timed out."
        }
    }
}

func requestSpeechAuthorization() throws {
    let semaphore = DispatchSemaphore(value: 0)
    var status: SFSpeechRecognizerAuthorizationStatus = .notDetermined

    SFSpeechRecognizer.requestAuthorization { authorizationStatus in
        status = authorizationStatus
        semaphore.signal()
    }

    _ = semaphore.wait(timeout: .now() + 10)

    guard status == .authorized else {
        throw CliError.authorizationDenied
    }
}

func transcribeAudioFile(at path: String, localeIdentifier: String) throws -> String {
    guard FileManager.default.fileExists(atPath: path) else {
        throw CliError.fileNotFound(path)
    }

    guard let recognizer = SFSpeechRecognizer(locale: Locale(identifier: localeIdentifier)),
          recognizer.isAvailable else {
        throw CliError.recognizerUnavailable(localeIdentifier)
    }

    let request = SFSpeechURLRecognitionRequest(url: URL(fileURLWithPath: path))
    request.shouldReportPartialResults = false

    let semaphore = DispatchSemaphore(value: 0)
    var transcript = ""
    var recognitionError: Error?

    let task = recognizer.recognitionTask(with: request) { result, error in
        if let result {
            transcript = result.bestTranscription.formattedString
            if result.isFinal {
                semaphore.signal()
            }
            return
        }

        if let error {
            recognitionError = error
            semaphore.signal()
        }
    }

    let waitResult = semaphore.wait(timeout: .now() + 120)
    task.cancel()

    if waitResult == .timedOut {
        throw CliError.timedOut
    }

    if let recognitionError {
        throw recognitionError
    }

    let trimmedTranscript = transcript.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmedTranscript.isEmpty else {
        throw CliError.emptyTranscript
    }

    return trimmedTranscript
}

@main
struct MacOSTranscribeCLI {
    static func main() {
        do {
            guard CommandLine.arguments.count >= 2 else {
                throw CliError.usage
            }

            let audioPath = CommandLine.arguments[1]
            let localeIdentifier = CommandLine.arguments.count >= 3 ? CommandLine.arguments[2] : "ko-KR"

            try requestSpeechAuthorization()
            let transcript = try transcribeAudioFile(at: audioPath, localeIdentifier: localeIdentifier)
            print(transcript)
        } catch {
            fputs("\(error.localizedDescription)\n", stderr)
            exit(1)
        }
    }
}
